src/backend/win_cid/file_dialog/dialog_ffi.rs

```diff
@@ -1,223 +1,478 @@
+use super::super::utils::str_to_vec_u16;
 use crate::FileDialog;
 
-use std::{ffi::OsStr, iter::once, os::windows::ffi::OsStrExt, path::PathBuf};
+use std::{ffi::c_void, path::PathBuf};
 
-use windows::core::{Result, PCWSTR, PWSTR};
-use windows::Win32::{
+use windows_sys::core::{GUID, HRESULT, PCWSTR, PWSTR};
+use windows_sys::Win32::{
     Foundation::HWND,
     System::Com::{CoCreateInstance, CoTaskMemFree, CLSCTX_INPROC_SERVER},
     UI::Shell::{
-        Common::COMDLG_FILTERSPEC, FileOpenDialog, FileSaveDialog, IFileDialog, IFileOpenDialog,
-        IFileSaveDialog, IShellItem, SHCreateItemFromParsingName, FILEOPENDIALOGOPTIONS,
-        FOS_ALLOWMULTISELECT, FOS_PICKFOLDERS, SIGDN_FILESYSPATH,
+        Common::COMDLG_FILTERSPEC, FileOpenDialog, FileSaveDialog, SHCreateItemFromParsingName,
+        FILEOPENDIALOGOPTIONS, FOS_ALLOWMULTISELECT, FOS_PICKFOLDERS, SIGDN, SIGDN_FILESYSPATH,
     },
 };
 
 use raw_window_handle::RawWindowHandle;
 
-unsafe fn read_to_string(ptr: PWSTR) -> String {
-    let mut len = 0usize;
+#[inline]
+unsafe fn read_to_string(ptr: *const u16) -> String {
     let mut cursor = ptr;
+
     loop {
-        let val = cursor.0.read();
-        if val == 0 {
+        if *cursor == 0 {
             break;
         }
-        len += 1;
-        cursor = PWSTR(cursor.0.add(1));
+
+        cursor = cursor.add(1);
     }
 
-    let slice = std::slice::from_raw_parts(ptr.0, len);
+    let slice = std::slice::from_raw_parts(ptr, cursor.offset_from(ptr) as usize);
     String::from_utf16(slice).unwrap()
 }
 
-pub enum DialogKind {
+pub type Result<T> = std::result::Result<T, HRESULT>;
+
+#[inline]
+fn wrap_err(hresult: HRESULT) -> Result<()> {
+    if hresult >= 0 {
+        Ok(())
+    } else {
+        Err(hresult)
+    }
+}
+
+#[repr(C)]
```

Comment:
This is the beginning of the COM bindings, let me know if you want them moved to a separate file for clarity. But basically each vtable just has the fields that are actually used in this file, the other ones are just replaced with usize to match the layout requirements without taking the compile time hit for the various parameters etc that aren't needed.

Reply:
Different module for sure.

Also, I'm not sure if I like the idea of using usize, I don't expect contributors to touch this madness every time a new method is needed, and tbh neither I want to review those. Preferably, it would just be dropped into a file, and we could just pretend that those don't exist :grimacing:, perhaps even just copy-pasted 1 to 1 from windows-rs

Reply:
Uhh, sure I can add the fully defined vtables.